### PR TITLE
refactor(docs,theme): split DocItem comp, useDoc hook

### PR DESCRIPTION
**Pull request by slorber**
## Breaking change

- `@theme/DocItem` component has been split into much smaller subcomponents
- `@theme/DocItemFooter` has been moved to `@theme/DocItem/Footer` and does not receive props anymore
- You can access previous props data with the new `useDoc()` hook

## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Motivation

This is a follow-up of moving in the direction to:
- using context instead of props => more flexible
- split large component into smaller ones 

We already did that in multiple places as well (doc sidebar etc...)

This PR does:
- put doc route data in the context
- access data through a new `useDoc()` hook, that will likely become part of our public API
- define that public API more explicitly (ie we don't have to stick to the shape in which data is injected into comps => implementation detail)
- split components into much smaller subcomponents, easier to swizzle
- make it easy to change the doc item layout without swizzling a large file

We'll likely mark some of these components as safe to swizzle later?

## Test Plan

Preview works as before

### Test links

Deploy preview: https://deploy-preview-7644--docusaurus-2.netlify.app/docs

## Related issues/PRs

Somehow related: https://github.com/facebook/docusaurus/discussions/7637 (useDoc() will make this easy)

### CODE REVIEW

File: packages/docusaurus-theme-classic/src/theme/DocItem/Layout/index.tsx
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/DocItem/Layout/index.tsx:25]: maybe we'll want to extract this hook later?

Not 100% satisfied with it atm

File: packages/docusaurus-theme-common/src/contexts/doc.tsx
Comment (slorber) [inline packages/docusaurus-theme-common/src/contexts/doc.tsx:12]: not sure how to name this context/hook/file

`useDoc` doesn't look too bad and is relatively simple

File: packages/docusaurus-theme-common/src/contexts/doc.tsx
Comment (slorber) [inline packages/docusaurus-theme-common/src/contexts/doc.tsx:25]: My opinion: storing metadata/frontmatter in the mdx file is an implementation detail. Consuming the data with `Component.metadata` doesn't feel very natural to me so I'd rather not stick to it 🤷‍♂️ 

In practice, we could create separate bundles/props for these metadata. And it may have advantages to do so in the future if for example we want to provide some kind of "data query layer" (I mean, for example we want to display some docs/blog metadata in the homepage, without having to load all the mdx content in that page)

I'm not sure how things will turn out, but I'd prefer to decouple the metadata from the MDX component when consuming it. I'd like to do something similar on the blog as well.

**Review by slorber — COMMENTED** [review]
Waiting for a review before merge

File: packages/docusaurus-theme-classic/src/theme-classic.d.ts
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/theme-classic.d.ts:288]: Since we are making breaking changes anyway, can we make this `DocItem/Footer`?

File: packages/docusaurus-theme-classic/src/theme/DocItem/Layout/index.tsx
Comment (Josh-Cena) [reply packages/docusaurus-theme-classic/src/theme/DocItem/Layout/index.tsx:25]: I don't like this—seems like over-abstraction that hides its semantics.

File: packages/docusaurus-theme-classic/src/theme-classic.d.ts
Comment (slorber) [reply packages/docusaurus-theme-classic/src/theme-classic.d.ts:288]: Yes 👍 also thought about it

File: packages/docusaurus-theme-classic/src/theme/DocItem/Layout/index.tsx
Comment (slorber) [reply packages/docusaurus-theme-classic/src/theme/DocItem/Layout/index.tsx:25]: What would you suggest?

I'd like the layout component to be easy to edit visually, without the TOC technical details noise.

Maybe it's fine to keep this weird hook, but just keep it there so that users can understand it more easily? 🤷‍♂️ 

I don't like it either but I don't have much better to suggest without polluting the component

Comment (Josh-Cena) [reply packages/docusaurus-theme-classic/src/theme/DocItem/Layout/index.tsx:25]: Yes, when I'm swizzling, I'd personally like to gain _some_ understanding of what the hook is doing. This is especially important for non-TS users. This hook is congregating too many unrelated things. It's fine if it's in the same file and easily viewable, but extracting it to `theme-common` makes it much harder to work with.

File: packages/docusaurus-theme-common/src/contexts/doc.tsx
Comment (Josh-Cena) [reply packages/docusaurus-theme-common/src/contexts/doc.tsx:25]: I agree—the metadata should be a separate chunk. This means we can easily provide the same metadata as both props and context. Seems too complex in the short term, though.

File: packages/docusaurus-theme-common/src/contexts/doc.tsx
Comment (Josh-Cena) [reply packages/docusaurus-theme-common/src/contexts/doc.tsx:12]: I think it's fine. Need more JSDoc on how you are supposed to use this

File: packages/docusaurus-theme-common/src/contexts/doc.tsx
Comment (Josh-Cena) [inline packages/docusaurus-theme-common/src/contexts/doc.tsx:56]: Looks incomplete? Also this is highly likely user-facing, so we can benefit from being more explicit about use-cases, since it's a quite lightweight wrapper around the existing prop.

File: packages/docusaurus-theme-classic/src/theme/DocItem/Layout/index.tsx
Comment (slorber) [reply packages/docusaurus-theme-classic/src/theme/DocItem/Layout/index.tsx:25]: agree 👍 

let's ship this as is and see how to improve later

File: packages/docusaurus-theme-common/src/contexts/doc.tsx
Comment (Josh-Cena) [inline packages/docusaurus-theme-common/src/contexts/doc.tsx:33]: Do we want to expose this as context? Looks like TMI to me 😄

Comment (slorber) [reply packages/docusaurus-theme-common/src/contexts/doc.tsx:33]: That's part of the static data being available in the subtree, so not sure how it could be used, but why not?

Eventually we may move all this subtree context data to be made available at the plugin level? Using our secret `useRouteContext`?

https://github.com/facebook/docusaurus/issues/6885

You'd prefer to forward a `children` prop? 

I think both solutions are fine 🤷‍♂️ it's unlikely someone would want to render that component in a weird place like the Footer, so forwarding `children` could also work

Comment (Josh-Cena) [reply packages/docusaurus-theme-common/src/contexts/doc.tsx:33]: I was thinking this is useless in all cases to be exposed as a context, since only the main container would care about it and it doesn't really encourage any code reuse. (e.g. no one would be forwarding this to some subcomponent as props.)

Comment (slorber) [reply packages/docusaurus-theme-common/src/contexts/doc.tsx:33]: Up to you, I'm fine if you revert it to `children` if you think it's better, both look fine to me

File: packages/docusaurus-theme-common/src/contexts/doc.tsx
Comment (Josh-Cena) [inline packages/docusaurus-theme-common/src/contexts/doc.tsx:38]: Are we intentionally cherry-picking the properties from `content` on both type and value, instead of using a spread?

Comment (slorber) [reply packages/docusaurus-theme-common/src/contexts/doc.tsx:38]: yes because I'd like to avoid mdx loader export implementation details to automatically leak into an API that we'll likely make public later

Comment (Josh-Cena) [reply packages/docusaurus-theme-common/src/contexts/doc.tsx:38]: Makes sense. Just confirming

**Review by Josh-Cena — APPROVED** [review]

**Comment by lanegoolsby** [comment]
@slorber is `useDoc` exported? I tried to import it in my component but get an error:

```
import React, { useState } from 'react';
import TOC from '@theme-original/TOC';
import { useDoc } from '@docusaurus/theme-common/src/contexts/doc';

export default function TOCWrapper(props: any) {
  const docMeta = useDoc();
  const [tags, setTags] = useState<any[]>([]);

  function asdf() {
    if (docMeta.frontMatter.tags && docMeta.frontMatter.tags.length > 0) {
      setTags(docMeta.frontMatter.tags);
    }
  }

  return (
    <>
      <TOC {...props} />
      Hello world
    </>
  );
}
```

```
Module not found: Error: Package path ./src/contexts/doc is not exported from package /my/repo/path/node_modules/@docusaurus/theme-common (see exports field in /my/repo/path/node_modules/@docusaurus/theme-common/package.json)
```

**Comment by TasoOneAsia** [comment]
@lanegoolsby Was looking into utilizing this hook as well, and from preliminary investigation seems like the useDoc hook is only being exported under the `internal` entry point for the time being.  Should be able to import it by referencing that sub-path instead, i.e.

```js
import { useDoc } from '@docusaurus/theme-common/internal'
```

**Comment by lanegoolsby** [comment]
Typescript and/or VSCode doesn't like that

image

**Comment by Josh-Cena** [comment]
@lanegoolsby You need to compile with `moduleResolution: NodeNext`.

**Comment by lanegoolsby** [comment]
Can this hook be used in a custom plugin? I can use it in a swizzled TOC, but I get this error when I try to use the same code in a custom plugin.

image

I am using a `/docs` path - /docs/tutorial-basics/create-a-document

I tried both with and without a `<DocProvider>`, both throw the same error.

```
import React, { useState } from 'react';
import { useDoc } from '@docusaurus/theme-common/internal';
import TOC from '@theme-init/TOC';
import { DocProvider } from '@docusaurus/theme-common/internal';

export default function (props: any) {
  const docMeta = useDoc();

  return (
    <DocProvider content={props.content}>
      <TOC {...props} />
      Hello world <code>{JSON.stringify(docMeta)}</code>
    </DocProvider>
  );
}
```

**Comment by slorber** [comment]
`useDoc()` is only supposed to be used on a doc page. Other pages (blog, pages or your custom plugin routes) do not have the same data available so you'd have to use another hook, eventually creating your own.

I'm not sure to understand at all what you are trying to do here 😅 please show a repro.

Note technically you could reuse `DocProvider` + `useDoc` (a React context is just a way to provide a value to a whole subtree), but these comp/hook are created with TS typings for the data made available through the docs plugin, so if you are looking to pass a very different set of data you'd rather create your own React context instead.

Note: you can't call the `useXyz()` hook in the same component that uses the provider `<ProvideXyz>`. What you do above is always wrong. You can only use the hook on components in the provider subtree. You can generally solve this problem by extracting what's inside the provider to a new custom React component, and using the hook there.

**Comment by lanegoolsby** [comment]
I'm trying to setup a custom plugin that injects a custom React component immediately below the Table of Contents component, for example:

image

Where the Hello World is, I'd like to have a component that will read all the tags for a doc page, query an internal API with said tags, and render the results in a grid.

**Comment by slorber** [comment]
> I'm trying to setup a custom plugin that injects a custom React component immediately below the Table of Contents component

Note the TOC component is more a "design system component" and it's not really meant to be coupled to a specific usage (docs, blog, pages). I mean it's not supposed to use `useDoc()` or `useBlog()`. Now we don't prevent it but I think you'd rather avoid creating such coupling.

If you only want to enhance the Doc TOC (and not the page/blog TOC), you should rather swizzle 
`@theme/DocItem/TOC/Desktop` (instead of `@theme/TOC`)  => this component can safely use `useDoc()`

Also don't forget to think about the mobile experience where the TOC is a collapsible at the top of the document

**Pull request merged**
